Persist targetSoc/Energy, minSoc, targetTime per vehicle, add minSoc to UI

naltatis opened the pull request:
Vehicle based settings now survive server restart and are carried over when manually changing vehicles at a loadpoint.

fixes #5271
replaces #6740

- 💽 storing `targetSoc`, `targetEnergy`, `minSoc`, `targetTime` per vehicle in database
- 🔄 stored settings are used on startup, vehicle change (UI) and physical vehicle (dis-)connect
- 🧪 removed "experimental" flag for `minSoc` UI (see screenshot)
  - 🛫 target time charging is now in the departure tab (default)
  - 🛬 additional onIdentify options (min/maxCurrent, mode) will be added to the arrival tab later
- 🎯 Charging plan link is now shown in all modes modes and with missing target energy (guest/offline vehicles)

Note: [`onIdentify`](https://docs.evcc.io/docs/reference/configuration/vehicles/#onidentify) values in `evcc.yaml` have precedent. From now on they should only be used if you want to reset to these values. @VolkerK62 could you updated the documentation accordingly?

shows active `minSoc`
![Bildschirmfoto 2023-05-21 um 11 42 36](https://github.com/evcc-io/evcc/assets/152287/82cde920-f1a1-458b-91f2-3243913aee01)

departure tab
![Bildschirmfoto 2023-05-21 um 11 43 01](https://github.com/evcc-io/evcc/assets/152287/e49a5fd4-9033-4760-a89d-7f3d24af4529)

arrival tab
![Bildschirmfoto 2023-05-21 um 12 28 32](https://github.com/evcc-io/evcc/assets/152287/21a0a7ba-2e60-40de-86e5-5a3918941e0b)

andig committed:
Plugins: make javascript return values more permissive

naltatis committed:
Persist targetSoc/Energy, minSoc and target time, remove experimental from minSoc UI

naltatis committed:
Merge branch 'master' into fix/persist_target_soc_time

naltatis committed:
go generate

naltatis committed:
Fix tests

naltatis committed:
Update minSoc description

andig commented on api/api.go:
```diff
@@ -152,6 +152,7 @@ type Vehicle interface {
 	Battery
 	BatteryCapacity
 	IconDescriber
+	Id() string
```
Das gehört da nicht rein. Die Fahrzeuge sollten nicht wissen, dass sie in einem Slice hängen. Ich hab eine Idee wie wir das umgehen könnten.

naltatis replied:
Ist wieder entfernt.

andig commented on core/loadpoint_vehicle.go:
```diff
@@ -218,6 +220,45 @@ func (lp *Loadpoint) publishVehicleFeature(f api.Feature) {
 }
 
+// persistVehicleSettings stores user configuration (via UI/API) for the current vehicle
+func (lp *Loadpoint) persistVehicleSettings() {
```
Ich würde versuchen das in den Coordinator zu schieben. Dann brauchts auch keine Id mehr.

naltatis replied:
Ich hab versucht das Persistieren komplett in den Corrdinator zu packen. Das führt aber zu Problemen, da der Coordinator öffentliche Methoden vom LP aufrufen muss (bspw. Set/GetTargetSoc) und wir dann in Locking Probleme laufen.

Ich hab jetzt die Id Logik entfernt. Stattdessen ermittelt der Coordinator den Index der Fahrzeuges. Auf der Basis wird gespeichert und geladen.

naltatis replied:
@andig Hast du hier noch Kommentare oder passt das für dich? Sonst würd ich das gerne reinholen.

andig commented on core/loadpoint_vehicle.go:
```diff
@@ -218,6 +220,45 @@ func (lp *Loadpoint) publishVehicleFeature(f api.Feature) {
+}
+
+// restoreVehicleSettings restores user configuration (via UI/API) for the current vehicle
+func (lp *Loadpoint) restoreVehicleSettings() {
```
Dito Coordinator

andig commented on go.mod:
```diff
@@ -104,6 +104,8 @@ require (
 	nhooyr.io/websocket v1.8.7
 )
 
+require github.com/gosimple/unidecode v1.0.1 // indirect
```
Da ist die Blockstruktur kaputt

naltatis replied:
Ist wieder raus.

andig commented on vehicle/embed.go:
```diff
@@ -57,3 +58,8 @@ var _ api.FeatureDescriber = (*embed)(nil)
 	return v.Features_
 }
+
+func (v *embed) Id() string {
```
Mhhhm. Liegt also gar nicht am Slice. Lass mich nochmal drüber nachdenken.

naltatis replied:
Ich hatte das erst mit Coordinator und Index, aber gerade im Hinblick auf die Config-UI find ich es ne gute Idee hier schon auf eine eindeutige ID zu gehen. Wenn ein Fahrzeug aus der DB kommt muss es seine ID sowieso kennen.

andig replied:
Nö, dem Fahrzeug ist das egal. Es weiss nicht, dass es gespeichert wird!

naltatis replied:
Siehe oben.

naltatis committed:
Store settings by index. Show plan outside pv mode. Add help texts.

naltatis committed:
remove slug

naltatis committed:
go mock

naltatis committed:
lint

naltatis committed:
Merge branch 'master' into fix/persist_target_soc_time

naltatis committed:
Merge branch 'master' into fix/persist_target_soc_time

andig committed:
chore: simplify templates (#8304)

naltatis committed:
Add e2e tests with playwright (#8123)

andig committed:
Revert "1p3p: let charger handle session stop/restart (#7723)"

This reverts commit dd787ce52fe07e617a73061fbb31393335f381dd.

C64Axel committed:
mazda2mqtt: document vin required (#8319)

andig committed:
FoxESS: split H1/H3 devices (#7376)

Co-authored-by: premultiply <4681172+premultiply@users.noreply.github.com>

weblate committed:
Translations update from Hosted Weblate (#8124)

* Translated using Weblate (Czech)

Currently translated at 37.3% (99 of 265 strings)

Co-authored-by: Dusan Suja <bc.suja.dusan@googlemail.com>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/cs/
Translation: evcc/evcc

* Translated using Weblate (Finnish)

Currently translated at 100.0% (265 of 265 strings)

Co-authored-by: Arna Lepikkö <arna.lepikko@telemail.fi>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/fi/
Translation: evcc/evcc

* Translated using Weblate (German)

Currently translated at 100.0% (265 of 265 strings)

Co-authored-by: ThinkEV <claas@rootdir.de>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/de/
Translation: evcc/evcc

* Translated using Weblate (Croatian)

Currently translated at 100.0% (265 of 265 strings)

Translated using Weblate (Slovenian)

Currently translated at 100.0% (265 of 265 strings)

Co-authored-by: Žiga Deisinger <ziga@deisinger.si>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/hr/
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/sl/
Translation: evcc/evcc

* Translated using Weblate (Catalan)

Currently translated at 100.0% (265 of 265 strings)

Co-authored-by: Norbert Poch <npoch@yahoo.com>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/ca/
Translation: evcc/evcc

* Translated using Weblate (Dutch)

Currently translated at 76.6% (203 of 265 strings)

Translated using Weblate (Dutch)

Currently translated at 74.3% (197 of 265 strings)

Co-authored-by: Ruben Van Boxem <vanboxem.ruben@gmail.com>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/nl/
Translation: evcc/evcc

* fix toml

---------

Co-authored-by: Dusan Suja <bc.suja.dusan@googlemail.com>
Co-authored-by: Arna Lepikkö <arna.lepikko@telemail.fi>
Co-authored-by: ThinkEV <claas@rootdir.de>
Co-authored-by: Žiga Deisinger <ziga@deisinger.si>
Co-authored-by: Norbert Poch <npoch@yahoo.com>
Co-authored-by: Ruben Van Boxem <vanboxem.ruben@gmail.com>
Co-authored-by: premultiply <4681172+premultiply@users.noreply.github.com>

andig committed:
Add OBO Betterman Ion (#8321)

Also aligns heartbeat implementations

salz3n committed:
Enphase: add token auth (firmware D7.x.xxx) and grid (#8247)

paschdan committed:
Mqtt: fix smartCostLimit topic case (#8328)

DerAndereAndi committed:
Add ISO15118 vehicle template (#8302)

andig committed:
Mqtt: simplify setters

andig committed:
Mqtt: disable message ordering to improve performance

andig committed:
chore: simplify random state generation

andig committed:
Porsche: remove deprecated mobile api (#8349)

andig committed:
Porsche: remove remaining mobile api types

lex777777 committed:
Cupra: add odometer (#8340)

andig committed:
chore: refactor go-stylish

andig committed:
chore: fix stale handler

C64Axel committed:
mazda2mqtt: longer timeout (#8364)

andig committed:
chore: pre-process toml

andig committed:
chore: prevent toml double quotes

andig committed:
Audi: temporarily switch to etron (#8374)

OAltr committed:
Fix nightly build (#8384)

premultiply committed:
Update SunSpec templates (#8270)

andig committed:
Revert "Fix nightly build (#8384)"

This reverts commit 536dbc936c7ced16f8bd00409a8cc00e5e807e57.

andig committed:
Revert "Add e2e tests with playwright (#8123)"

This reverts commit 0df815729b5471bc7c448857d60137684b926611.

GrimmiMeloni committed:
Easee: update Observation IDs (#8391)

GrimmiMeloni committed:
Easee: handle smartCharging errors (#8389)

RTTTC committed:
TWC: add non-Tesla vehicle warning (#8329)

andig committed:
TWC: allow loadpoint to wakeup vehicle (#8284)

andig committed:
chore: use gridx consts

DerAndereAndi committed:
Fix ISO15118 vehicle (#8402)

Such vehicles would be reported as `(Offline)` previously, as no Soc can be fetched and thus needs to behave as the `offline` template.

The SoC is fetched via the charger, so hardcode it to 0 in here

kscholty committed:
Modbus: add coils  (#8385)

andig committed:
OpenEVSE: fix api (#8415)

andig committed:
chore: refactor offline vehicles (#8404)

GrimmiMeloni committed:
Easee: wait for api confirmation (#8307)

andig committed:
Revert "Easee: wait for api confirmation (#8307)"

This reverts commit ce0e4277aa9b9f7b58c12d7bf0958ae7f632c729.

nalxnet committed:
automatically switch session log energy unit (#8371)

Show the charged energy in the session log table and details in Wh for
values below 1 kWh and in kWh for larger values

naltatis committed:
Fix nightly/release + integration workflow  (#8450)

* Revert "Add e2e tests with playwright (#8123)"

This reverts commit 0df815729b5471bc7c448857d60137684b926611.

* Revert "Revert "Add e2e tests with playwright (#8123)""

This reverts commit 2299306954045dfe62541d5520a66b5ab7ec4971.

* fix nightly/release

* fix nightly/release

* fix nightly/release

---------

Co-authored-by: andig <cpuidle@gmx.de>

naltatis committed:
chore: no integration on nightly

naltatis committed:
chore: no integration on nightly

naltatis committed:
chore: no integration on nightly

naltatis committed:
Sessions: filter, monthly sums, paging (#8278)

naltatis committed:
fmt

andig commented:
@naltatis könntest Du rebasen? Für mich ist schwer zu erkennen, was hier PR und was Beiwerk ist.

andig commented:
Die Lösung, die ID vom Coordinator zu holen find ich übrigens Klasse, ist aber nicht ganz stabil. Akzeptieren wir fürs Erste?

naltatis commented:
>  ist aber nicht ganz stabil. Akzeptieren wir fürs Erste? 

Ja ich würd sagen lass uns damit starten und auf "echte IDs" umschalten sobald wir das haben. Ich sehe das hier in der Tat auch eher als einen Zwischenschritt.

naltatis committed:
Merge branch 'master' into fix/persist_target_soc_time

naltatis commented:
> Imho ist lokaler rebase weniger für solche Fehler anfällig.

Ich find ja mergen weniger fehleranfällig. Aber seis drum. Der Branch ist nun auf Stand und das Diff sieht sauber aus. Schau gerne noch mal.

savus4 commented on i18n/en.toml:
```diff
@@ -169,7 +169,7 @@ label = "Max. Current"
 
 [main.loadpointSettings.minSoc]
-description = "For emergencies. The vehicle gets „fast” charged to {0}% from all available solar, and then continues with only the solar surplus."
+description = "The vehicle gets „fast” charged to {0}% in solar mode. Then continues with solar surplus. Usefull to ensure a minimum range even for darker days."
```
Grad nur zufällig gesehen :)
```suggestion
description = "The vehicle gets „fast” charged to {0}% in solar mode. Then continues with solar surplus. Useful to ensure a minimum range even for darker days."
```

andig committed:
Update i18n/en.toml

Co-authored-by: Simon Riepl <43091717+savus4@users.noreply.github.com>

andig committed:
Use exp/slices

andig committed:
Use mux consistently

naltatis committed:
Merge branch 'master' into fix/persist_target_soc_time

JanDragon commented:
Is there a planned date to merge this PR? I would be especially interested in the feature that allows you to set the vehicle to always be at a certain charge level. Should be the minSOC option. Looking forward to use that in the future.  Thanks in advance.

naltatis committed:
Merge branch 'master' into fix/persist_target_soc_time

naltatis committed:
add integration tests

Robbe64 commented:
Ich teste das Feature gerade im Nightly aber bei mir vergisst er `targetSoc` und `minSoc` beim ab und wieder anstecken. Soll ich die Details hier hinterlegen oder einen Bug anlegen?

naltatis commented:
Hast du die Werte aus der yaml entfernt? siehe #9245

Robbe64 commented:
Ja, hab sie entfernt. Starte ich EVCC neu, bleiben die Werte die ich eingestellt hatte erhalten. Auch beim Fahrzeugwechsel in EVCC bleiben die Werte bestehen. Nur wenn ich ab und anstecke klappt es nicht. Dann wird `targetSoc` auf 100% und `minSoc` auf 0/x gestellt. 

```yaml
#uri: 0.0.0.0:7070 # uri for ui
plant: *****
interval: 20s # control cycle interval

log: debug
levels:
  id: trace
  easee: trace
#  lp-1: trace
#  tibber: trace

sponsortoken: *****

meters:
  - type: template
    template: kostal-plenticore
    id: 71
    host: 192.168.178.101
    port: 1502
    usage: pv
    modbus: tcpip
    name: pv1
  - type: template
    template: kostal-plenticore
    id: 71
    host: 192.168.178.101
    port: 1502
    usage: battery
    modbus: tcpip
    name: battery2
    capacity: 5.2
  - type: template
    template: kostal-ksem-inverter
    id: 71
    host: 192.168.178.101
    port: 1502
    usage: grid
    modbus: tcpip
    name: grid3

chargers:
  - name: easee1
    type: easee
    user: *****
    password: *****
    charger: EHE84FNG
    timeout: 20s

loadpoints:
  - title: Carport
    charger: easee1
    mode: pv
    mincurrent: 7
    maxcurrent: 32
    resetOnDisconnect: true
    soc:
      estimate: false # Interpolation des SoC zwischen den Abfragen zum Auto
      poll:
        mode: always
        interval: 60m
    vehicle: egolf
    vehicles:
      - egolf
#      - tesla
    enable:
      delay: 3m
    disable:
      delay: 7m
      threshold: 100
    guardduration: 5m
    priority: 10

site:
  title: Mein Zuhause
  meters:
    grid: grid3
    pvs:
      - pv1
    batteries:
      - battery2
  bufferSoC: 95 # Hausbatterie wird oberhalb SoC x% als Puffer genutzt
  prioritySoC: 20 # Hausbatterie bekommt bis zum SoC x% Priorität beim laden
  residualPower: 100 # 100W soll immer eingespeist werden. Nötig für Batteriesystem
  maxGridSupplyWhileBatteryCharging: 50 # Verhindert den Netzbezug bei DC Leistung > AC Leistung und gleichzeitigem Batterieladen

vehicles:
  - name: egolf
    type: vw
    title: eGolf
    capacity: 21
    user: *****
    password: *****
    phases: 1
    onIdentify:
      minCurrent: 7
      maxCurrent: 16
#      targetSoC: 80 #-> Now stored from UI
    #cloud: false
    identifiers: 04DA682ABD7280

tariffs:
  currency: EUR
  grid:
    type: tibber
    token: *****
    #price: 0.31 # [currency]/kWh

  feedin:
    # rate for feeding excess (pv) energy to the grid
    type: fixed
    price: 0.077 # [currency]/kWh
  co2:
    type: grünstromindex
    zip: 59609

influx:
  url: http://solaranzeige:8086
  database: evcc

telemetry: true
```

Robbe64 commented:
Habe gerade festgestellt, dass schon beim Abstecken die genannten Werte zurückgesetzt werden. Womöglich beim setzen des Standardfahrzeugs?
```
[site  ] DEBUG 2023/08/07 11:48:36 ----
[lp-1  ] DEBUG 2023/08/07 11:48:36 charge power: 0W
[site  ] DEBUG 2023/08/07 11:48:37 pv power: 912W
[site  ] DEBUG 2023/08/07 11:48:37 battery soc: 39%
[site  ] DEBUG 2023/08/07 11:48:37 battery power: -515W
[site  ] DEBUG 2023/08/07 11:48:37 grid power: -31W
[site  ] DEBUG 2023/08/07 11:48:37 site power: -446W
[lp-1  ] DEBUG 2023/08/07 11:48:37 charge currents: [0 0 0]A
[lp-1  ] DEBUG 2023/08/07 11:48:37 charge total import: 4443.962kWh
[lp-1  ] DEBUG 2023/08/07 11:48:37 charger status: A
[lp-1  ] INFO 2023/08/07 11:48:37 car disconnected
[lp-1  ] DEBUG 2023/08/07 11:48:37 set charge mode: pv
[lp-1  ] DEBUG 2023/08/07 11:48:37 set min current: 7
[lp-1  ] DEBUG 2023/08/07 11:48:37 set max current: 32
[lp-1  ] DEBUG 2023/08/07 11:48:37 set min soc: 0
[lp-1  ] DEBUG 2023/08/07 11:48:37 set target soc: 100
[lp-1  ] DEBUG 2023/08/07 11:48:37 set priority: 10
[lp-1  ] DEBUG 2023/08/07 11:48:37 set min current: 7
[lp-1  ] DEBUG 2023/08/07 11:48:37 set max current: 16
[lp-1  ] DEBUG 2023/08/07 11:48:37 vehicle soc: 83%
[lp-1  ] DEBUG 2023/08/07 11:48:37 vehicle range: 93km
[site  ] DEBUG 2023/08/07 11:48:37 ----
```

naltatis commented:
Ok, klingt dann doch nach nem Bug. Magst du ein Issue aufmachen?

Robbe64 commented:
Habe #9283 angelegt
